pkg/framework/spec_unmarshal.go

```diff
@@ -0,0 +1,83 @@
+package framework
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/stackrox/helmtest/internal/parser"
+	yamlv3 "gopkg.in/yaml.v3"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	filenamePragma = `#!helmtest-filename:`
+)
+
+// injectFilename stores the filename into a "fake" foot comment of the respective node. This allows us to reconstruct
+// the filename in UnmarshalYAML, even though it isn't stored in the node itself (and there is no context that would
+// allow us to pass other values).
+func injectFilename(node *yamlv3.Node, filename string) {
+	node.FootComment += fmt.Sprintf("\n%s%s", filenamePragma, filename)
+	for _, child := range node.Content {
+		injectFilename(child, filename)
+	}
+}
+
+// UnmarshalYAML unmarshals a test spec from YAML, preserving line information that we're interested in.
+func (t *Test) UnmarshalYAML(node *yamlv3.Node) error {
+	// Create an alias of this type without a custom UnmarshalYAML method.
+	type testNoMethods Test
+	if err := node.Decode((*testNoMethods)(t)); err != nil {
+		return err
+	}
+
+	if node.Kind != yamlv3.MappingNode {
+		return nil // weird but ok
+	}
+
+	// We've moved the filename into a fake footer comment, to allow extracting it here.
+	var filename string
+	for _, line := range strings.Split(node.FootComment, "\n") {
+		line = strings.TrimSpace(line)
+		if strings.HasPrefix(line, filenamePragma) {
+			filename = line[len(filenamePragma):]
+			break
+		}
+	}
+
+	// Check for mapping keys of interest.
+	for i := 0; i < len(node.Content); i += 2 {
+		valueNode := node.Content[i+1]
+		valueSrcContext := parser.SourceContext{
+			Filename: filename,
+			Line:     valueNode.Line - 1, // Node.Line is one-based, but SourceContext.Line is zero-based
+		}
+		// In literal and folded style, the value only begins on the preceding line.
+		if valueNode.Style == yamlv3.LiteralStyle || valueNode.Style == yamlv3.FoldedStyle {
+			valueSrcContext.Line++
+		}
+
+		keyNode := node.Content[i]
+		if keyNode.Value == "defs" {
+			t.defsSrcCtx = valueSrcContext
+		} else if keyNode.Value == "expect" {
+			t.expectSrcCtx = valueSrcContext
+		}
+	}
+	return nil
+}
+
+// UnmarshalYAML unmarshals a RawDict from YAML, making sure that the resulting type matches the result of
+// json.Unmarshal on the equivalent JSON.
+func (d *RawDict) UnmarshalYAML(node *yamlv3.Node) error {
+	yamlBytes, err := yamlv3.Marshal(node)
+	if err != nil {
+		return err
+	}
+	jsonBytes, err := yaml.YAMLToJSON(yamlBytes)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(jsonBytes, d)
```

Comment:
Can you remind me why it was necessary to convert into JSON?

Reply:
Helm itself uses `sigs.k8s.io/yaml` for YAML marshaling/unmarshaling, which is based on YAML-to-JSON. I think that converting to JSON is generally preferred because JSON and YAML are regarded as merely alternative forms of representation. Therefore, you generally do not want more than one `struct` tag (`json`, `yaml`) to determine your representation, and since JSON is the more restricted case, it makes sense to standardize on that. That's how the types in this library were written; I only added `yaml` struct tags out of necessity because I needed the node parsing interface.

As to why we need to convert to JSON here: apparently the `gopkg.in/yaml` library and the JSON-based YAML deserializers differ in subtle ways when unmarshaling into a `map[string]interface{}`. Therefore, we take the YAML `Node` representation, convert it to YAML (there's no JSON-serialization defined for it), convert that YAML to JSON, and unmarshal JSON. It's a lot of back-and-forth, but this is the only way I could think of to ensure that we can use the YAML `Node` representation without sacrificing JSON unmarshaling semantics.

Reply:
Thanks for the explanation!